### ISSUE TITLE
indexing for separable sums

### ISSUE THREAD

**Issue by mehrhardt** [odlgroup/odl#897]
I want to create a functional f that is the sum over a lot of f_i s. So I code something like:
```python
f = list()
for i in range(len(Xn)):
    f.append(1./2 * odl.solvers.L2NormSquared(X).translated(-data[i]))
    
sumf = odl.solvers.functional.SeparableSum(*f)
```
However, in my application it would be most elegant to call `f = sumf` such that both calls
`f(x)` and `f[i](x[i])` make sense. I have seen that ODL supports `f.functionals[i](x[i])` which would be ok but not really elegant.

Moreover, it would be good if the syntax is invariant under some formulations. For instance
```python
f = list()
for i in range(len(Xn)):
    f.append(odl.solvers.L2NormSquared(X).translated(-data[i]))
    
sumf = 1./2 * odl.solvers.functional.SeparableSum(*f)
```
would not support `f.functionals[i](x[i])` any more but needs `f.functional.functionals[i](x[i])`.

Edit: @adler-j improved typesetting

**Comment by adler-j** [comment]
Regarding the first part, I made a pull request with the suggested syntax, #898.

Regarding the second, I see what you mean, but that is a much harder suggestion to implement, given that we would need to add some intelligence, i.e. that scalar multiplication commutes with separable sums. I guess it would be possible, but we really have to consider if it is worth it in terms of added complexity.

Any thoughts @kohr-h?

**Comment by kohr-h** [comment]
Regarding the "intelligence" issue, my view is that while
```python
scaled_sum = scalar * odl.solvers.SeparableSum(*funcs)
```
and
```python
scaled_funcs = [scalar * func for func in funcs]
scaled_sum = odl.solvers.SeparableSum(*scaled_funcs)
```
*compute* the same, the *are not* the same. 
- With the first version, you express the intention to do the scaling outside, which is probably slightly more efficient. The result is no longer a `SeparableSum`, and rightly so IMO.
- The second version forces the scaling to happen inside, and by that you in some sense express the intention that you want the end result to be a `SeparableSum`. This happens at the expense of (probably negligible) extra computational cost and a slightly more verbose syntax.

I understand that you would like to have both ;-) but in this case I prefer not to put in extra intelligence. 
The reason is that it would then be hard to argue against other cases that are also "trivial" to simplify in a similar manner, which adds complexity often for a questionable gain.

So my suggestion for you @MatthiasJE would be to use the second syntax to scale the functionals individually, and then you have a `SeparableSum`, which will be indexable after #898 is merged.

**Comment by kohr-h** [comment]
> The reason is that it would then be hard to argue against other cases that are also "trivial" to simplify in a similar manner, which adds complexity often for a questionable gain.

A typical example of this would be the addition of two `SeparableSum` functionals, which is also trivially a `SeparableSum`. But to realize that, you need to start putting in checks like
```python
def __add__(self, other):
    if isinstance(other, SeparableSum):
        # add member functionals, return SeparableSum(*sums)
    else:
        # do other stuff
```
which makes the code more fragile.

**Comment by adler-j** [comment]
My suggestion is that we close this issue for now and make a larger overall issue w.r.t to adding more intelligence (i.e let ODL figure out that multiplication commutes with indexing) where we can have this discussion at a higher level and talk about how we want to handle this long term.

The "smaller things" have already been fixed by #898.

Thank you for the input @MatthiasJE!

**Comment by kohr-h** [comment]
I'll close this now since I don't see anything we can do without fundamentally changing the way functionals behave.
